### ISSUE TITLE
Capture issues when using roam internal functions in `org-roam-find-file-hook` 

### ISSUE THREAD

**Issue by adql** [org-roam/org-roam#1678]
### Description

`org-roam-find-file-hook` is run whether an existing file is visited or a capture process for a new file begins. I want to use this hook to conditionally make some visual changes in only certain types of notes (I use tags beginning with `@` to differentiate classical "Zettel" from notes referring to e.g. `@book` and `@article`).

However, calling something like `(org-roam-node-tags (org-roam-node-at-point))` from this hook yields

> or: Wrong type argument: org-roam-node, nil

So even if one only wants to add a hook after a visit, it is called also in capture but before there is proper access to the `node` struct. This poses a limit on what can be done with that hook.

My solution is to include only org-roam files _and_ exclude org-roam-capture buffers by writing the function as

```elisp
(when (org-roam-file-p)
    (unless (or org-roam-capture--node
    ...
    ))
```
and then add the hook to both `org-roam-find-file-hook` and `org-capture-after-finalize-hook`. This does have the desired effects in the buffer after capture has been completed and when it is visited later. But this is somewhat cumbersome, and it also doesn't play well with `org-roam-capture-new-node-hook` for having the same effects immediately in capture.

#### Steps to Reproduce

1. `(add-hook 'org-roam-find-file-hook (lambda () org-roam-node-tags (org-roam-node-at-point)))`
2. `(org-roam-node-find)`
3. Create a new note.

#### Backtrace

```elisp
Debugger entered--Lisp error: (wrong-type-argument org-roam-node nil)
  signal(wrong-type-argument (org-roam-node nil))
  (or (and (memq (type-of cl-x) cl-struct-org-roam-node-tags) t) (signal 'wrong-type-argument (list 'org-roam-node cl-x)))
  (progn (or (and (memq (type-of cl-x) cl-struct-org-roam-node-tags) t) (signal 'wrong-type-argument (list 'org-roam-node cl-x))) (aref cl-x 15))
  (let ((cl-x (org-roam-node-at-point))) (progn (or (and (memq (type-of cl-x) cl-struct-org-roam-node-tags) t) (signal 'wrong-type-argument (list 'org-roam-node cl-x))) (aref cl-x 15)))
  (closure (t) nil (let ((cl-x (org-roam-node-at-point))) (progn (or (and (memq (type-of cl-x) cl-struct-org-roam-node-tags) t) (signal 'wrong-type-argument (list 'org-roam-node cl-x))) (aref cl-x 15))))()
  run-hooks(org-roam-find-file-hook)
  (progn (run-hooks 'org-roam-find-file-hook))
  (if (org-roam-file-p) (progn (run-hooks 'org-roam-find-file-hook)))
  org-roam--file-setup()
  run-hooks(find-file-hook)
  after-find-file(t t)
  find-file-noselect-1(#<buffer 20210726103453-test.org> "~/org-roam-test/20210726103453-test.org" nil nil "~/org-roam-test/20210726103453-test.org" nil)
  find-file-noselect("/home/amir/org-roam-test/20210726103453-test.org")
  org-capture-target-buffer("/home/amir/org-roam-test/20210726103453-test.org")
  (set-buffer (org-capture-target-buffer path))
  (let ((head x42) (path x40)) (setq path (expand-file-name (string-trim (org-roam-capture--fill-template path t)) org-roam-directory)) (set-buffer (org-capture-target-buffer path)) (if (file-exists-p path) nil (org-roam-capture--put :new-file path) (insert (org-roam-capture--fill-template head t))) (widen) (setq p (goto-char (point-min))))
  (if (null x43) (let ((head x42) (path x40)) (setq path (expand-file-name (string-trim (org-roam-capture--fill-template path t)) org-roam-directory)) (set-buffer (org-capture-target-buffer path)) (if (file-exists-p path) nil (org-roam-capture--put :new-file path) (insert (org-roam-capture--fill-template head t))) (widen) (setq p (goto-char (point-min)))) nil)
  (let* ((x42 (car x41)) (x43 (cdr x41))) (if (null x43) (let ((head x42) (path x40)) (setq path (expand-file-name (string-trim (org-roam-capture--fill-template path t)) org-roam-directory)) (set-buffer (org-capture-target-buffer path)) (if (file-exists-p path) nil (org-roam-capture--put :new-file path) (insert (org-roam-capture--fill-template head t))) (widen) (setq p (goto-char (point-min)))) nil))
  (if (consp x41) (let* ((x42 (car x41)) (x43 (cdr x41))) (if (null x43) (let ((head x42) (path x40)) (setq path (expand-file-name (string-trim (org-roam-capture--fill-template path t)) org-roam-directory)) (set-buffer (org-capture-target-buffer path)) (if (file-exists-p path) nil (org-roam-capture--put :new-file path) (insert (org-roam-capture--fill-template head t))) (widen) (setq p (goto-char (point-min)))) nil)) nil)
  (let* ((x40 (car x39)) (x41 (cdr x39))) (if (consp x41) (let* ((x42 (car x41)) (x43 (cdr x41))) (if (null x43) (let ((head x42) (path x40)) (setq path (expand-file-name (string-trim ...) org-roam-directory)) (set-buffer (org-capture-target-buffer path)) (if (file-exists-p path) nil (org-roam-capture--put :new-file path) (insert (org-roam-capture--fill-template head t))) (widen) (setq p (goto-char (point-min)))) nil)) nil))
  (if (consp x39) (let* ((x40 (car x39)) (x41 (cdr x39))) (if (consp x41) (let* ((x42 (car x41)) (x43 (cdr x41))) (if (null x43) (let ((head x42) (path x40)) (setq path (expand-file-name ... org-roam-directory)) (set-buffer (org-capture-target-buffer path)) (if (file-exists-p path) nil (org-roam-capture--put :new-file path) (insert ...)) (widen) (setq p (goto-char ...))) nil)) nil)) nil)
  (let* ((x39 (cdr val))) (if (consp x39) (let* ((x40 (car x39)) (x41 (cdr x39))) (if (consp x41) (let* ((x42 (car x41)) (x43 (cdr x41))) (if (null x43) (let (... ...) (setq path ...) (set-buffer ...) (if ... nil ... ...) (widen) (setq p ...)) nil)) nil)) nil))
  (cond ((eq x30 'file) (let* ((x31 (cdr val))) (if (consp x31) (let* ((x32 (car x31)) (x33 (cdr x31))) (if (null x33) (let (...) (setq path ...) (if ... nil ...) (set-buffer ...) (widen) (setq p ...)) nil)) nil))) ((eq x30 'file+olp) (let* ((x34 (cdr val))) (if (consp x34) (let* ((x35 (car x34)) (x36 (cdr x34))) (if (consp x36) (let* (... ...) (if ... ... nil)) nil)) nil))) ((eq x30 'file+head) (let* ((x39 (cdr val))) (if (consp x39) (let* ((x40 (car x39)) (x41 (cdr x39))) (if (consp x41) (let* (... ...) (if ... ... nil)) nil)) nil))) ((eq x30 'file+head+olp) (let* ((x44 (cdr val))) (if (consp x44) (let* ((x45 (car x44)) (x46 (cdr x44))) (if (consp x46) (let* (... ...) (if ... ... nil)) nil)) nil))) ((eq x30 'file+datetree) (let* ((x51 (cdr val))) (if (consp x51) (let* ((x52 (car x51)) (x53 (cdr x51))) (if (consp x53) (let* (... ...) (if ... ... nil)) nil)) nil))) ((eq x30 'node) (let* ((x56 (cdr val))) (if (consp x56) (let* ((x57 (car x56)) (x58 (cdr x56))) (if (null x58) (let (...) (let ... ... ... ...)) nil)) nil))) (t nil))
  (let* ((x30 (car val))) (cond ((eq x30 'file) (let* ((x31 (cdr val))) (if (consp x31) (let* ((x32 ...) (x33 ...)) (if (null x33) (let ... ... ... ... ... ...) nil)) nil))) ((eq x30 'file+olp) (let* ((x34 (cdr val))) (if (consp x34) (let* ((x35 ...) (x36 ...)) (if (consp x36) (let* ... ...) nil)) nil))) ((eq x30 'file+head) (let* ((x39 (cdr val))) (if (consp x39) (let* ((x40 ...) (x41 ...)) (if (consp x41) (let* ... ...) nil)) nil))) ((eq x30 'file+head+olp) (let* ((x44 (cdr val))) (if (consp x44) (let* ((x45 ...) (x46 ...)) (if (consp x46) (let* ... ...) nil)) nil))) ((eq x30 'file+datetree) (let* ((x51 (cdr val))) (if (consp x51) (let* ((x52 ...) (x53 ...)) (if (consp x53) (let* ... ...) nil)) nil))) ((eq x30 'node) (let* ((x56 (cdr val))) (if (consp x56) (let* ((x57 ...) (x58 ...)) (if (null x58) (let ... ...) nil)) nil))) (t nil)))
  (if (consp val) (let* ((x30 (car val))) (cond ((eq x30 'file) (let* ((x31 (cdr val))) (if (consp x31) (let* (... ...) (if ... ... nil)) nil))) ((eq x30 'file+olp) (let* ((x34 (cdr val))) (if (consp x34) (let* (... ...) (if ... ... nil)) nil))) ((eq x30 'file+head) (let* ((x39 (cdr val))) (if (consp x39) (let* (... ...) (if ... ... nil)) nil))) ((eq x30 'file+head+olp) (let* ((x44 (cdr val))) (if (consp x44) (let* (... ...) (if ... ... nil)) nil))) ((eq x30 'file+datetree) (let* ((x51 (cdr val))) (if (consp x51) (let* (... ...) (if ... ... nil)) nil))) ((eq x30 'node) (let* ((x56 (cdr val))) (if (consp x56) (let* (... ...) (if ... ... nil)) nil))) (t nil))) nil)
  (let* ((val (or (org-roam-capture--get :if-new) (user-error "Template needs to specify `:if-new'")))) (if (consp val) (let* ((x30 (car val))) (cond ((eq x30 'file) (let* ((x31 ...)) (if (consp x31) (let* ... ...) nil))) ((eq x30 'file+olp) (let* ((x34 ...)) (if (consp x34) (let* ... ...) nil))) ((eq x30 'file+head) (let* ((x39 ...)) (if (consp x39) (let* ... ...) nil))) ((eq x30 'file+head+olp) (let* ((x44 ...)) (if (consp x44) (let* ... ...) nil))) ((eq x30 'file+datetree) (let* ((x51 ...)) (if (consp x51) (let* ... ...) nil))) ((eq x30 'node) (let* ((x56 ...)) (if (consp x56) (let* ... ...) nil))) (t nil))) nil))
  (let (p) (let* ((val (or (org-roam-capture--get :if-new) (user-error "Template needs to specify `:if-new'")))) (if (consp val) (let* ((x30 (car val))) (cond ((eq x30 'file) (let* (...) (if ... ... nil))) ((eq x30 'file+olp) (let* (...) (if ... ... nil))) ((eq x30 'file+head) (let* (...) (if ... ... nil))) ((eq x30 'file+head+olp) (let* (...) (if ... ... nil))) ((eq x30 'file+datetree) (let* (...) (if ... ... nil))) ((eq x30 'node) (let* (...) (if ... ... nil))) (t nil))) nil)) (prog1 (save-excursion (goto-char p) (let* ((node (and t org-roam-capture--node)) (id (and node (org-roam-node-id node)))) (if id (org-entry-put p "ID" id))) (prog1 (org-id-get-create) (run-hooks 'org-roam-capture-new-node-hook))) (org-roam-capture--adjust-point-for-capture-type)))
  org-roam-capture--goto-location()
  (cond ((plist-get org-roam-capture--info :ref) (let* ((node (and t (org-roam-capture--get-node-from-ref (plist-get org-roam-capture--info :ref))))) (if node (progn (set-buffer (org-capture-target-buffer (org-roam-node-file node))) (goto-char (org-roam-node-point node)) (widen) (org-end-of-subtree t t)) (org-roam-capture--goto-location)))) ((and (org-roam-node-file org-roam-capture--node) (org-roam-node-point org-roam-capture--node)) (set-buffer (org-capture-target-buffer (org-roam-node-file org-roam-capture--node))) (goto-char (org-roam-node-point org-roam-capture--node)) (widen) (org-end-of-subtree t t) (org-roam-node-id org-roam-capture--node)) (t (org-roam-capture--goto-location)))
  (let ((id (cond ((plist-get org-roam-capture--info :ref) (let* ((node ...)) (if node (progn ... ... ... ...) (org-roam-capture--goto-location)))) ((and (org-roam-node-file org-roam-capture--node) (org-roam-node-point org-roam-capture--node)) (set-buffer (org-capture-target-buffer (org-roam-node-file org-roam-capture--node))) (goto-char (org-roam-node-point org-roam-capture--node)) (widen) (org-end-of-subtree t t) (org-roam-node-id org-roam-capture--node)) (t (org-roam-capture--goto-location))))) (org-capture-put :template (org-roam-capture--fill-template (org-capture-get :template))) (org-roam-capture--put :id id) (org-roam-capture--put :finalize (or (org-capture-get :finalize) (org-roam-capture--get :finalize))))
  org-roam-capture--get-point()
  org-capture-set-target-location(nil)
  org-capture(nil nil)
  (let* ((props (plist-put props :call-location (point-marker))) (org-capture-templates (mapcar #'(lambda (template) (org-roam-capture--convert-template template props)) (or templates org-roam-capture-templates))) (org-roam-capture--node node) (org-roam-capture--info info)) (if (and (not keys) (= (length org-capture-templates) 1)) (progn (setq keys (car (car org-capture-templates))))) (org-capture goto keys))
  (progn (let ((--cl-keys-- --cl-rest--)) (while --cl-keys-- (cond ((memq (car --cl-keys--) '(:goto :keys :node :info :props :templates :allow-other-keys)) (setq --cl-keys-- (cdr (cdr --cl-keys--)))) ((car (cdr (memq ... --cl-rest--))) (setq --cl-keys-- nil)) (t (error "Keyword argument %s not one of (:goto :keys :node ..." (car --cl-keys--)))))) (let* ((props (plist-put props :call-location (point-marker))) (org-capture-templates (mapcar #'(lambda (template) (org-roam-capture--convert-template template props)) (or templates org-roam-capture-templates))) (org-roam-capture--node node) (org-roam-capture--info info)) (if (and (not keys) (= (length org-capture-templates) 1)) (progn (setq keys (car (car org-capture-templates))))) (org-capture goto keys)))
  (let* ((goto (car (cdr (plist-member --cl-rest-- ':goto)))) (keys (car (cdr (plist-member --cl-rest-- ':keys)))) (node (car (cdr (plist-member --cl-rest-- ':node)))) (info (car (cdr (plist-member --cl-rest-- ':info)))) (props (car (cdr (plist-member --cl-rest-- ':props)))) (templates (car (cdr (plist-member --cl-rest-- ':templates))))) (progn (let ((--cl-keys-- --cl-rest--)) (while --cl-keys-- (cond ((memq (car --cl-keys--) '...) (setq --cl-keys-- (cdr ...))) ((car (cdr ...)) (setq --cl-keys-- nil)) (t (error "Keyword argument %s not one of (:goto :keys :node ..." (car --cl-keys--)))))) (let* ((props (plist-put props :call-location (point-marker))) (org-capture-templates (mapcar #'(lambda ... ...) (or templates org-roam-capture-templates))) (org-roam-capture--node node) (org-roam-capture--info info)) (if (and (not keys) (= (length org-capture-templates) 1)) (progn (setq keys (car (car org-capture-templates))))) (org-capture goto keys))))
  org-roam-capture-(:node #s(org-roam-node :file nil :file-hash nil :file-atime nil :file-mtime nil :id nil :level nil :point nil :todo nil :priority nil :scheduled nil :deadline nil :title "test" :properties nil :olp nil :tags nil :aliases nil :refs nil) :props (:finalize find-file :call-location #<marker at 2142 in 20-roam.el>))
  (if (progn (or (and (memq (type-of node) cl-struct-org-roam-node-tags) t) (signal 'wrong-type-argument (list 'org-roam-node node))) (aref node 1)) (org-roam-node-visit node other-window) (org-roam-capture- :node node :props '(:finalize find-file :call-location #<marker at 2142 in 20-roam.el>)))
  (let ((node (org-roam-node-read initial-input filter-fn))) (if (progn (or (and (memq (type-of node) cl-struct-org-roam-node-tags) t) (signal 'wrong-type-argument (list 'org-roam-node node))) (aref node 1)) (org-roam-node-visit node other-window) (org-roam-capture- :node node :props '(:finalize find-file :call-location #<marker at 2142 in 20-roam.el>))))
  org-roam-node-find()
  funcall-interactively(org-roam-node-find)
  call-interactively(org-roam-node-find nil nil)
  command-execute(org-roam-node-find)
```

#### Expected Results

Hooks added to `org-roam-find-file-hook` work in the capture buffer.

#### Actual Results

> or: Wrong type argument: org-roam-node, nil

### Environment

- Emacs: GNU Emacs 27.2 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.27, cairo version 1.17.4)
 of 2021-03-26
- Framework: N/A
- Org: Org mode version 9.4.6 (9.4.6-11-g1ee52c-elpaplus @ /home/amir/.emacs.d/elpa/org-plus-contrib-20210719/)
- Org-roam: 2.0.0

- Org-roam commit: https://github.com/jethrokuan/org-roam/commit/6dc8dda

**Comment by adql** [comment]
And of course - thanks for the great work!

**Comment by jethrokuan** [comment]
`org-roam-find-file-hook` is meant to be run each time an Org-roam file is visited, and this does include capture buffers. It sets up completions, db updating which are equally important for new capture buffers. So I do think the current implementation is accurate.

Your hook seems to be dependent on whether the node has been cached into the database, and you seem to only want it to run once, on `find-file`. I'm not sure if there's a good way to support that at all tbh...

**Comment by adql** [comment]
I see what you mean, this hook is used heavily for internal processes and considering the workflow of the database it's not a bug.

My motivation is that when working the Luhmann/Ahrens way there are different "boxes" for literature ref. notes and for proper networked notes. In my case "literature" includes any notes referring to some resource and are tagged `@article`, `@book`, `@podcast` etc. I also have notes tagged `@topic` which together form some sort of an index. I use the templates to add these in capture. All these literature and meta notes should look differently from proper Zettelkasten notes – with the latter being displayed in a more writing-oriented way (distraction free, larger line spacing, etc.)

In my current setting I just excluded the capture buffer from testing for note tags, which works fine by additionally adding the function to `org-capture-after-finalize-hook` as I wrote above. It does mean though that the visuals are not modified during capture, but I'm currently fine with that. I suppose a solution could be to use the templates to file the @-notes in sub-directories and then test for those, since they don't require the db.
